### PR TITLE
Add confirm button component

### PR DESCRIPTION
**Pull request by EreMaijala**
First of all, this is very much related to #2723, but I believe this could benefit the work in #2723 and make it easier to transition to the new JS that doesn't use Bootstrap dropdown.

The main purpose here is to move all confirmation dropdowns into a component that works just like before but needs less markup in templates and provides a single place for the implementation. Since we have a few different ways of using a confirm button (as a button, as a submit button, as a link), the implementation is a bit more complex that I'd like, but the result is that not many other changes were needed, and all Mink tests pass without modifications (apart from one exception that relied on specific DOM structure).

One more design aspect: I tried to hide as much of the implementation as possible, so the parameters to the component reflect that. E.g. classes are optional, and e.g. an icon link should just work without additional classes. This should make it easier to change the underlying implementation without issues, but there certainly still has to be some awareness of the styling and markup.

### CODE REVIEW

**Comment by EreMaijala** [comment]
@sturkel89 Could you verify that all confirmations still work as they should? Affected functions:

- Deletion of records from cart
- Emptying cart
- Purging of items from checkout history (both buttons)
- Canceling of holds, interlibrary loan requests and storage retrieval requests (both buttons)
- Deletion of a library card
- Renewal of loans (both buttons)
- Removal of favorites from a list or favorites

You should be able to test all these with the Demo driver, but you'll need to enable some settings in the Catalog section of config.ini.

File: module/VuFind/src/VuFind/View/Helper/Root/Component.php [hunk 1]
```diff
@@ -52,6 +52,8 @@ class Component extends AbstractHelper
      */
     public function __invoke(string $name, $params = []): string
     {
+        static $invokation = 0;
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/Component.php]: Nitpick: should probably be spelled "invocation".

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @EreMaijala -- I didn't study all the refactoring in close detail (I trust @sturkel89's testing more than my eyes for that), but the new component template makes sense to me, and I didn't see any major problems at a glance. Just one spelling nitpick...

**Review by sturkel89 — COMMENTED** [review]
I tested most of the cases described in the PR. I'm not sure if I'm supposed to be reporting on the behavior of the **dropdown** or the presence or absence of a **banner** indicating whether the requested action was successful, so I will report on both.

At the bottom of this review I've put a summary of the tweaks and changes I would suggest to improve consistency across the various places the dropdown and confirmation come into play. I don't know that any of them are dealbreakers, so I will submit this review as a comment, not a Request Changes. Thanks!!

### BOOK BAG
**Item removal**
I can select items, click the Delete option, and see a **Yes/No dropdown**.  (Why no sentence asking "Do you wish to remove these item(s) from your book bag?)
image

If I select Yes, the items disappear, but there is **no confirmation banner**.

**Empty bookbag**
With items in the book bag, I can click Empty Book Bag (without checking any individual boxes or selecting any entries) and the **Yes/No dropdown** appears. After I click Yes, I get a blue-background pop-up confirmation:

image

-----------------------

### HOLDS, ILL REQUESTS, and STORAGE RETRIEVAL REQUESTS
**Item removal**
In these three situations, the dropdown for item selection asks a confirming question: **"Do you wish to cancel your selected X?" Yes/No**

image

In all cases, clicking Yes results in a green-background banner across the top indicating that the request was successful, and/or a red-background banner indicating any unsuccessful cancellations. (The unsuccessful ones are indicated below the item as well.)

image

image

**Cancel all of these**
When you ask to cancel all holds/ILL requests/storage retrieval requests, you are shown a dropdown with a confirming sentence and Yes/No options:
image

If the results are that everything is cancelled, there's a banner across the top saying "X request(s) were successfully canceled".  I had one case where there were two items, and one was successfully cancelled and another was not. In that case, I only got a green banner across the top but NO RED BANNER AT THE TOP - the only red banner appeared under the relevant item:

image

This behavior is different from an experience I had with mixed results with holds and recalls when I had selected several individual items and tried to cancel selected holds:

image

Maybe cancelling SELECTED holds shows all the results at the top of the screen, but cancelling ALL holds only shows a banner for the successful cancellations?

---------------------

### LOAN RENEWAL
**Selected items:**
I checked the box for one loan and clicked the box for Renew Selected Items.  I got a dropdown with a sentence "Do you wish to renew the selected items?" Yes/No

image

Successful renewals are ONLY indicated by a green bar at the item - no green bar across the top of the screen.

image

Unsuccessful renewals went through a few different phases. It seems that the first failure is only marked below the record with a red banner, with no banner at the top of the screen:
image

A repeated attempt sometimes is met with a red banner at the top saying "simulated account block" and/or a blue banner saying "Simulated account block; try again and it will work eventually."

image

I'm not sure if there is a pattern to this sequence. A lot of behavior in Demo seems a bit random, so this might not be a clear diagnostic.

**Renew all items**
When I tried to renew all items, I got a similar “Do you wish to renew all your items?” question with a Yes/No dropdown. After clicking yes, the failure notifications seemed to be at the top of the screen, not at the individual items. The success notifications are at the items, NOT at the top of the screen.

I guess it makes sense to have successful renewal notices at the items, since the items are still on the screen. (Cancelled requests eliminate the cancelled item from the list, so the confirmation banner must appear at the top of the screen.) However, I would like to see the green success banner across the top of the screen as well, for consistency.

--------------

### FAVORITES LISTS
**Item removal:**
I opened a favorites list, clicked the checkboxes for two items, and clicked the Delete button. A detailed popup appeared, showing the title of the list and the titles of the items I had marked for deletion. (Why not the “Do you wish to delete the selected saved items from the list?” and Yes/No?)

image

When I clicked Delete, I got this green-background confirmation popup:

image

**Deleting a favorites list:**
When I am viewing a favorites list and I click Delete List, I get a semi-detached Yes/No dropdown with no contextualizing sentence.
image

When I click Yes, I get a confirmation that the list has been deleted (although I see that the items are still in my Saved Items list).

image

---------

### LIBRARY CARDS
I clicked the delete button next to one library card, and got a floating Yes/No dropdown:
image

When I clicked Yes, the card disappeared and I got a green notification banner across the top of the screen:

image

------------------

I didn't know how to enable transaction history, so I haven't tested that yet.

----------

### Summary of suggested changes:

**_Bookbag:_**
- Add a confirming sentence to the Yes/No dropdown for both buttons.
- Add a confirmation banner after individual items are removed from the bookbag.

**_Holds/ILL/Storage Retrieval:_**
- Make the behavior consistent: always show green notification banners across the top when these requests are successfully cancelled (since there is no longer an item on the list) and red notification banners across both the top and at the item when there is a failure to cancel the hold/request.

This behavior should be the same whether the user selected individual items or said to cancel all of that type of request.

**_Loan renewal:_**
- Put all confirmations across the top of the screen for successful and unsuccessful renewals, in addition to the existing confirmations at the items.

**_Favorites list:_**
- Item deletion: This is set up differently from everything else. Can it be made to look more like the other features? There is no Yes/No option - just a list of items and one button to Delete (or the X outside the lightbox to cancel).

- Deleting an entire list: the Yes/No dropdown is floating instead of being located directly below the Delete List option. 

**_Library cards:_**
- The Yes/No dropdown is floating instead of being located directly below the Delete option.

**Comment by sturkel89** [comment]
Demian told me about Demo.ini, so I turned on TransactionHistory!

The dropdown behaves as usual:

image

The confirmation notice is BLUE instead of green!

image

Same for purging the entire loan history:

image

Not sure why some notifications of success are blue and others are green... I would have to put the entire list in one document and try to figure out if there is a logic to it. Without doing that, my thought is that every successful operation should have a green confirmation banner.

File: module/VuFind/src/VuFind/View/Helper/Root/Component.php (hunk 1 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/View/Helper/Root/Component.php]: Fixed!

**Comment by EreMaijala** [comment]
@sturkel89 I've intentionally avoided changing any functionality here, so the color of confirmations and lack of a header should be just like in dev branch. Likewise with the floating confirmations when the button is a link instead of a button (they're not actually floating, but they look like it because the links don't have borders). I agree that unifying things would make sense, but I'd leave that to (yet another) Jira ticket..

**Comment by sturkel89** [comment]
> I agree that unifying things would make sense, but I'd leave that to (yet another) Jira ticket..

Should that be separate Jira tickets for the different types of fixes, or one Jira ticket for all of this notification/confirmation related stuff (i.e., my summary list of suggested changes)? Does Jira support checkboxes for separate pieces of a request?

**Comment by EreMaijala** [comment]
> Should that be separate Jira tickets for the different types of fixes, or one Jira ticket for all of this notification/confirmation related stuff (i.e., my summary list of suggested changes)? Does Jira support checkboxes for separate pieces of a request?

For me grouping closely related things into one works fine and often makes sense. Sometimes it's difficult to judge, but issues can also be split later on. The important thing is to have the important observations there so that they don't get lost. :)

**Review by demiankatz — APPROVED** [review]
Thanks for fixing the spelling issue, @EreMaijala.

@sturkel89, if I understand correctly, all of the controls worked correctly during your testing, and your observations were about existing issues unrelated to the changes here. Is that correct? If so, I think we can go ahead and merge this -- but I'll wait until you confirm before doing so, in case I overlooked anything.

**Comment by sturkel89** [comment]
> @sturkel89, if I understand correctly, all of the controls worked correctly during your testing, and your observations were about existing issues unrelated to the changes here. Is that correct? If so, I think we can go ahead and merge this -- but I'll wait until you confirm before doing so, in case I overlooked anything.

Yes, the controls worked correctly. Good to merge!

**Pull request merged**